java/arcs/core/storage/StorageProxy.kt

```diff
@@ -172,13 +171,22 @@ class StorageProxy<Data : CrdtData, Op : CrdtOperationAtTime, T>(
 
                 if (applyFailures) {
                     notifyDesync()
-                    requestSynchronization()
-                    return true
+                    // Before we return, let's issue a request for synchronization on a new
+                    // coroutine. It can't be done on this current coroutine because the response
+                    // we give here is being waited-on downstream, and we could end up dead-locking
+                    // if requestSynchronization ends up needing a result from this to continue
+                    // (which is what happens with the StorageService).
+                    CoroutineScope(coroutineContext).launch { requestSynchronization() }
+                    return false
                 }
 
                 // all ops from storage applied cleanly so resolve waiting syncs
                 futuresToResolve.forEach { it.complete(value) }
-                notifyUpdate(message.operations)
+
+                // Notify our handles of an update if these operations came from elsewhere.
+                if (message.id != storeListenerId) {
```

Comment:
Does this mean if you write locally to the storageProxy it isn't going to notify updates? Won't this break when two particles have two handles which share the same proxy, one particle writing to a Handle won't trigger the onUpdate of the other?

Reply:
The method this is inside is only called by the Store. There's a separate "applyOp" from the handle-side that issues the updates to the other handles.